File: app/src/main/java/com/abishov/hexocat/home/trending/TrendingView.java
package com.abishov.hexocat.home.trending;

import android.content.Context;
import android.os.Build;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.widget.Toast;
import butterknife.BindDimen;
import butterknife.BindView;
import butterknife.ButterKnife;
import com.abishov.hexocat.R;
import com.abishov.hexocat.common.views.DividerItemDecoration;
import com.abishov.hexocat.home.repository.RepositoryAdapter;
import com.jakewharton.rxbinding2.support.v4.widget.RxSwipeRefreshLayout;
import com.jakewharton.rxbinding2.view.RxView;
import io.reactivex.Observable;
import io.reactivex.functions.Consumer;

public final class TrendingView extends FrameLayout {

  @BindView(R.id.swipe_refresh_layout_trending)
  SwipeRefreshLayout swipeRefreshLayout;

  @BindView(R.id.recyclerview_trending)
  RecyclerView recyclerViewTrending;

  @BindView(R.id.button_retry)
  Button buttonRetry;

  @BindView(R.id.textview_error)
  TextView textViewError;

  @BindDimen(R.dimen.trending_divider_padding_start)
  float dividerPaddingStart;

  private RepositoryAdapter repositoryAdapter;

  public TrendingView(Context context, @Nullable AttributeSet attrs) {
    super(context, attrs);
  }

  @Override
  protected void onFinishInflate() {
    super.onFinishInflate();
    ButterKnife.bind(this);

    repositoryAdapter = new RepositoryAdapter(LayoutInflater.from(getContext()),
        viewModel -> Toast.makeText(getContext(), viewModel.name(), Toast.LENGTH_SHORT).show());
    RecyclerView.LayoutManager recyclerViewLayoutManager = new LinearLayoutManager(getContext());
    recyclerViewTrending.setLayoutManager(recyclerViewLayoutManager);
    recyclerViewTrending.setAdapter(repositoryAdapter);
    recyclerViewTrending.addItemDecoration(new DividerItemDecoration(getContext(),
        DividerItemDecoration.VERTICAL, dividerPaddingStart, isRtl()));

    buttonRetry.setVisibility(View.GONE);
  }

  public Observable<Object> onSwipeRefreshLayout() {
    return RxSwipeRefreshLayout.refreshes(swipeRefreshLayout);
  }

  public Observable<Object> onRetry() {
    return RxView.clicks(buttonRetry);
  }

  public Consumer<TrendingViewState> bindTo() { // NOPMD
    return state -> {
      recyclerViewTrending.setVisibility(state.isSuccess() ? View.VISIBLE : View.GONE);
      swipeRefreshLayout.setRefreshing(state.isInProgress());
      buttonRetry.setVisibility(state.isFailure() ? View.VISIBLE : View.GONE);
      textViewError.setVisibility(state.isFailure() ? View.VISIBLE : View.GONE);

      if (state.isSuccess()) {
        repositoryAdapter.accept(state.items());
      } else if (state.isFailure()) {
        textViewError.setText(state.error());
      }
    };
  }

  private boolean isRtl() {
    return Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2 &&
        getLayoutDirection() == View.LAYOUT_DIRECTION_RTL;
  }
}

File: app/src/main/java/com/abishov/hexocat/home/trending/TrendingView.kt
package com.abishov.hexocat.home.trending

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import android.widget.FrameLayout
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import butterknife.BindView
import butterknife.ButterKnife
import com.abishov.hexocat.R
import com.abishov.hexocat.common.views.DividerItemDecoration
import com.abishov.hexocat.home.repository.RepositoryAdapter
import com.abishov.hexocat.home.repository.RepositoryAdapter.TrendingViewClickListener
import com.abishov.hexocat.home.repository.RepositoryViewModel
import com.jakewharton.rxbinding2.support.v4.widget.RxSwipeRefreshLayout
import com.jakewharton.rxbinding2.view.RxView
import io.reactivex.Observable

class TrendingView(context: Context, attrs: AttributeSet?) : FrameLayout(context, attrs) {

  @BindView(R.id.swipe_refresh_layout_trending)
  internal lateinit var swipeRefreshLayout: SwipeRefreshLayout

  @BindView(R.id.recyclerview_trending)
  internal lateinit var recyclerViewTrending: RecyclerView

  @BindView(R.id.button_retry)
  internal lateinit var buttonRetry: Button

  @BindView(R.id.textview_error)
  internal lateinit var textViewError: TextView

  private lateinit var repositoryAdapter: RepositoryAdapter

  private var dividerPaddingStart: Float =
    resources.getDimension(R.dimen.trending_divider_padding_start)

  private val isRtl: Boolean
    get() {
      return layoutDirection == View.LAYOUT_DIRECTION_RTL
    }

  override fun onFinishInflate() {
    super.onFinishInflate()
    ButterKnife.bind(this)

    buttonRetry.visibility = View.GONE

    repositoryAdapter = RepositoryAdapter(
      LayoutInflater.from(context),
      onRepositoryItemClick()
    )

    recyclerViewTrending.apply {
      layoutManager = LinearLayoutManager(context)
      adapter = repositoryAdapter

      addItemDecoration(
        DividerItemDecoration(context, DividerItemDecoration.VERTICAL, dividerPaddingStart, isRtl)
      )
    }
  }

  fun onSwipeRefreshLayout(): Observable<Any> {
    return RxSwipeRefreshLayout.refreshes(swipeRefreshLayout)
  }

  fun onRetry(): Observable<Any> {
    return RxView.clicks(buttonRetry)
  }

  fun bindTo(state: TrendingViewState) {
    recyclerViewTrending.visibility =
        if (state is TrendingViewState.Success) {
          View.VISIBLE
        } else {
          View.GONE
        }

    swipeRefreshLayout.isRefreshing = state is TrendingViewState.InProgress
    buttonRetry.visibility = if (state is TrendingViewState.Failure) View.VISIBLE else View.GONE
    textViewError.visibility = if (state is TrendingViewState.Failure) View.VISIBLE else View.GONE

    if (state is TrendingViewState.Success) {
      repositoryAdapter.accept(state.items)
    } else if (state is TrendingViewState.Failure) {
      textViewError.text = state.error
    }
  }

  private fun onRepositoryItemClick(): TrendingViewClickListener {
    return object : TrendingViewClickListener {
      override fun onRepositoryClick(repository: RepositoryViewModel) {
        context.startActivity(
          Intent(Intent.ACTION_VIEW, Uri.parse(repository.url))
        )
      }
    }
  }
}
